### ISSUE TITLE
Rework `canEdit` methods

### ISSUE THREAD

**Issue by luke-** [humhub/humhub#5987]
Currently there are two `$post->canEdit()` and `$post->content->canEdit()` implementations. 
We should reduce/rework this to one extendable solution.

**Comment by luke-** [comment]
Also related: https://github.com/humhub/humhub/issues/4594

**Comment by yurabakhtin** [comment]
@luke- PR https://github.com/humhub/humhub/pull/5992:

I have implemented to avoid the recursion what we had in the `tasks` module - https://github.com/humhub/tasks/pull/218/files#diff-5df20afb5542403995d9be9795c2e7dfdcac29bf86fc31979b78e12c9d81c03aL829.

---

Also there is a code comment: `// ToDo: Implement this as interface`.
We could do this with new interface `Editable` and method `canEdit(User $user): bool`, but please note after that we should update many external modules in order to avoid a compatiable error, because some of them use different implementation for the method `canEdit`, e.g. https://github.com/humhub/custom-pages/blob/master/models/CustomContentContainer.php#L294 - `humhub\modules\custom_pages\models\CustomContentContainer extends ContentActiveRecord` -> `public function canEdit($type = null): bool` - we have `$type` instead of `$user`.

**Comment by luke-** [comment]
@yurabakhtin Thanks for the PR. I think we should not only fix the problem and improve the code architecture here as well. This will mean some refactoring on the modules as well, unfortunately. 

Here are some thoughts:

- Do we really need two `canEdit()` methods? So one in the `ContentActiveRecord` and one in `Content`. Does it make sense to have it only in one class?

- Which methods like `canEdit` also make sense? e.g. `canDelete`, 'canArchive', ...? Possibly also `canAdd` (Your recent change with the `CreatePermission` could be related to that).

- An interface would make sense if we also want to call the same methods outside of `Content` context e.g. in Comments. Are there any cases other here? 

- Basically, central core classes like `Content` or `ContentActiveRecord` should be as compact as possible. Are there perhaps other approaches to implement the `can*` permission methods? e.g. `$post->permissions->canEdit()` and `$post->$permissionManager=StandardPermissionManager::class`.

**Comment by ArchBlood** [comment]
What about `canBlock` or `canCreate`?

**Comment by yurabakhtin** [comment]
@luke- 

> - Do we really need two `canEdit()` methods? So one in the `ContentActiveRecord` and one in `Content`. Does it make sense to have it only in one class?

Ideally we should not keep `canEdit()` method in the `ContentActiveRecord`, but as we can see on the `Task` model we have there additional conditions like `$this->hasTaskResponsible()` and `$this->isTaskResponsible()`, i.e. some external modules may have such additional checks, and they cannot be handled from `Content` side.

> - Which methods like `canEdit` / `canView` also make sense? e.g. `canDelete`, `canArchive`, ...?

Same as above I am not sure we can remove other methods like `canView` from classes extended from `ContentActiveRecord`.

> - Possibly also `canAdd` (Your recent change with the `CreatePermission` could be related to that).

I see the `CreatePermission` is used from `Content->canEdit()` depending on `$model->isNewRecord`.
Yes we can create a separate `canAdd` but I don't see a reason, because develop will needs to be sure it is a new record before use the `canAdd`, so it may occurs issues.

> - An interface would make sense if we also want to call the same methods outside of `Content` context e.g. in Comments. Are there any cases other here?

I don't find the `$model->canEdit()` is used from `Comment` or from `ContentAddonActiveRecord`.

> - Basically, central core classes like `Content` or `ContentActiveRecord` should be as compact as possible. Are there perhaps other approaches to implement the `can*` permission methods? e.g. `$post->permissions->canEdit()` and `$post->$permissionManager=StandardPermissionManager::class`.

Yes, we could try to create some new `StandardPermissionManager` and move into this new manager all `can*` from `Content`, so all `Content->can*()` methods should be deprecated in current version and deleted in some future version.
As I understand you mean something like `ContentContainerActiveRecord->$permissionManager`, right?

**Comment by luke-** [comment]
> 
> > * Do we really need two `canEdit()` methods? So one in the `ContentActiveRecord` and one in `Content`. Does it make sense to have it only in one class?
> 
> Ideally we should not keep `canEdit()` method in the `ContentActiveRecord`, but as we can see on the `Task` model we have there additional conditions like `$this->hasTaskResponsible()` and `$this->isTaskResponsible()`, i.e. some external modules may have such additional checks, and they cannot be handled from `Content` side.

That would then rather speak for having the `canEdit/*` methods once in the `ContentActiveRecord`, then they could also be extended there? A con would be that we bloat the class.  

> > * Which methods like `canEdit` / `canView` also make sense? e.g. `canDelete`, `canArchive`, ...?
> 
> Same as above I am not sure we can remove other methods like `canView` from classes extended from `ContentActiveRecord`.

I am not about removing them. When we redesign it, we should consider what cases there are and what additional methods we may need or need not.

> 
> > * Possibly also `canAdd` (Your recent change with the `CreatePermission` could be related to that).
> 
> I see the `CreatePermission` is used from `Content->canEdit()` depending on `$model->isNewRecord`. Yes we can create a separate `canAdd` but I don't see a reason, because develop will needs to be sure it is a new record before use the `canAdd`, so it may occurs issues.

Such a `Post::canAdd($contentContainer, [$user])` might look quite good. 

Might not be bad if we can remove complexity from the `canEdit` method and swap parts into a `canAdd`. e.g. https://github.com/humhub/humhub/blob/master/protected/humhub/modules/content/models/Content.php#L771

> 
> > * An interface would make sense if we also want to call the same methods outside of `Content` context e.g. in Comments. Are there any cases other here?
> 
> I don't find the `$model->canEdit()` is used from `Comment` or from `ContentAddonActiveRecord`.

We have a lot of `can*` methods here. Of course, it would be nice to find a consistent approach here that works for both (content & comments).

https://github.com/humhub/humhub/blob/master/protected/humhub/modules/content/components/ContentAddonActiveRecord.php#L123-L193

> 
> > * Basically, central core classes like `Content` or `ContentActiveRecord` should be as compact as possible. Are there perhaps other approaches to implement the `can*` permission methods? e.g. `$post->permissions->canEdit()` and `$post->$permissionManager=StandardPermissionManager::class`.
> 
> Yes, we could try to create some new `StandardPermissionManager` and move into this new manager all `can*` from `Content`, so all `Content->can*()` methods should be deprecated in current version and deleted in some future version. As I understand you mean something like `ContentContainerActiveRecord->$permissionManager`, right?

Yes, I was thinking of something similar to `PermissionManager`. Each content type would share a common default class and could override&extend it if needed. However, we need to be careful with the class naming here to avoid a mix up with the existing `PermissionManager` concept.

Currently, I prefer this approach. Also because we can isolate the logic nicely in its own class.

**Comment by yurabakhtin** [comment]
@luke-

> > > * Do we really need two `canEdit()` methods? So one in the `ContentActiveRecord` and one in `Content`. Does it make sense to have it only in one class?
> > 
> > 
> > Ideally we should not keep `canEdit()` method in the `ContentActiveRecord`, but as we can see on the `Task` model we have there additional conditions like `$this->hasTaskResponsible()` and `$this->isTaskResponsible()`, i.e. some external modules may have such additional checks, and they cannot be handled from `Content` side.
> 
> That would then rather speak for having the `canEdit/*` methods once in the `ContentActiveRecord`, then they could also be extended there? A con would be that we bloat the class.

Ok, as I understand you prefer to keep the `canEdit/*` methods only in the `ContentActiveRecord` and it would be better to use new `$permissionManager`.

> > > * Which methods like `canEdit` / `canView` also make sense? e.g. `canDelete`, `canArchive`, ...?
> > 
> > 
> > Same as above I am not sure we can remove other methods like `canView` from classes extended from `ContentActiveRecord`.
> 
> I am not about removing them. When we redesign it, we should consider what cases there are and what additional methods we may need or need not.

Ok.

> > > * Possibly also `canAdd` (Your recent change with the `CreatePermission` could be related to that).
> > 
> > 
> > I see the `CreatePermission` is used from `Content->canEdit()` depending on `$model->isNewRecord`. Yes we can create a separate `canAdd` but I don't see a reason, because develop will needs to be sure it is a new record before use the `canAdd`, so it may occurs issues.
> 
> Such a `Post::canAdd($contentContainer, [$user])` might look quite good.
> 
> Might not be bad if we can remove complexity from the `canEdit` method and swap parts into a `canAdd`. e.g. https://github.com/humhub/humhub/blob/master/protected/humhub/modules/content/models/Content.php#L771

Yes, we can create new `canAdd` indeed and swap the checks in order to do this for existing record firstly.

> > > * An interface would make sense if we also want to call the same methods outside of `Content` context e.g. in Comments. Are there any cases other here?
> > 
> > 
> > I don't find the `$model->canEdit()` is used from `Comment` or from `ContentAddonActiveRecord`.
> 
> We have a lot of `can*` methods here. Of course, it would be nice to find a consistent approach here that works for both (content & comments).
> 
> https://github.com/humhub/humhub/blob/master/protected/humhub/modules/content/components/ContentAddonActiveRecord.php#L123-L193

Previous time I thought you want to have an interface only for the method `canEdit` but if we need the interface for all `can*` methods then we should choose another name, not as `Editable`, however if we will use new `$permissionManager` then don't need to have the interface, right?

> > > * Basically, central core classes like `Content` or `ContentActiveRecord` should be as compact as possible. Are there perhaps other approaches to implement the `can*` permission methods? e.g. `$post->permissions->canEdit()` and `$post->$permissionManager=StandardPermissionManager::class`.
> > 
> > 
> > Yes, we could try to create some new `StandardPermissionManager` and move into this new manager all `can*` from `Content`, so all `Content->can*()` methods should be deprecated in current version and deleted in some future version. As I understand you mean something like `ContentContainerActiveRecord->$permissionManager`, right?
> 
> Yes, I was thinking of something similar to `PermissionManager`. Each content type would share a common default class and could override&extend it if needed. However, we need to be careful with the class naming here to avoid a mix up with the existing `PermissionManager` concept.
> 
> Currently, I prefer this approach. Also because we can isolate the logic nicely in its own class.

Ok, we can use a different name as you suggested `StandardPermissionManager` or `ContentPermissionManager`.

**Comment by yurabakhtin** [comment]
From discussion:

- Another idea: What if we really try to use the existing permission manager. the main problem here is, that the current implementation always uses Groups. So we could change `PermissionManager` > `ContentContainerPermissionManager` to `AbstractPermissionManager` > `GroupPermissionManager` > `ContentContainerPermissionManager`
- Then implement a new `AbstractPermissionManager` > `ContentPermissionManager`
- This would produces calls like `$post->permissionManager->can(CanEdit)`. If e.g. Task can somehow overwrite the `canEdit` class
- One big problem is here, that all Permissions are currently inheritet from `\humhub\libs\BasePermission` which would be better named `GroupPermission`
- we add a new default class `humhub\modules\content\permissions\CanEdit` including all logic
- Task could create `humhub\modules\tasks\permissions\CanEdit`
- I am not sure it is correctly to use `->can(CanEdit::class)` because from core side we cannot know that for the external model `Task` we should use also external permission, for example, we may need this to build a dropdown menu for wall entry. I.e. better to keep some `ContentPermissionManager` per each external model like Task, and call only `$post->permissionManager->canEdit()`. I mean on the `humhub\modules\content\widgets\EditLink` we call `$this->model->content->canEdit()` before to display the link edit
- e.g. we could introduce an attribute in the `ContentActiveRecord::$canEditPermission="humhub\modules\content\permissions\CanEdit"`. To check we could use: `$task->permissions->can($this->canEditPermission)`. Since `->permissions` is the new `ContentPermissionManager` it also could provide a shortcut called `canEdit()`?
- `EditLink` must now call `$this->model->permissions->canEdit()` this calls internally `$this->permissions->can($model->canEditPermission)`

**Comment by luke-** [comment]
@yurabakhtin Thanks, these should be the most important points.

**Comment by yurabakhtin** [comment]
@luke- I have started the new `ContentPermissionManager` in the PR https://github.com/humhub/humhub/pull/6010.

> One big problem is here, that all Permissions are currently inheritet from `\humhub\libs\BasePermission` which would be better named `GroupPermission`

Do you think we really need this renaming? So all external modules should be updated and their min humhub version should be `1.14`?

**Comment by yurabakhtin** [comment]
@luke- Please look at the method `humhub\modules\content\components\ContentPermissionManager->can($permission)`.
Almost all code there was copied from the `Content->canEdit()`.
You can find there the code `$this->model->content->container->getPermissionManager($user)->can($permission)`.

So new `ContentPermissionManager` looks like a wrapper for the original `PermissionManager` of the container.

I am not sure we should keep the new `AbstractPermissionManager` because cannot find what common methods or properties really shuuld be used and in original container/group `PermissionManager` and in the new `ContentPermissionManager`.

It seems we need the `ContentPermissionManager` only to move all methods `can*` from `Content`, because each method is such specific, so I don't see how it can be implemented through single method `ContentPermissionManager->can($permission)`.

**Comment by yurabakhtin** [comment]
@luke- Commit https://github.com/humhub/humhub/pull/6010/commits/365b7f17992c09df7cebfbeac295ef91c4750cd8: I have moved the method `Content->canMove()` to `ContentPermissionManager->canMove()`.

**Comment by yurabakhtin** [comment]
@luke- I have moved all methods `can*` from `Content` into `ContentPermissionManager`.

**Comment by luke-** [comment]
@yurabakhtin Can you take a look at my PR / Approach?

I tried to outsource the single permissions (canEdit/canView) into own permission classes and stay as close as possible to the existing PermissionManager concept.

https://github.com/humhub/humhub/pull/6034

**Comment by luke-** [comment]
> @luke- Yes, your aproach looks better, thank you for displaying what you wanted to see.
> 
> * I have done a small fix here [2223c4a](https://github.com/humhub/humhub/commit/2223c4a9aa7594cd6736aafc8bdf5505dec9573f) because current `$user` was null so all edit and view links were hidden.

Thanks!

> * I think we could remove `$content` the first param from the method `AbstractContentPermission->verify(ContentActiveRecord $content, ?User $user)` and initialize this as property when we call this method from the `ContentPermissionManager->verify(BasePermission $permission)`, like code:
> 
> ```
> if ($permission instanceof AbstractContentPermission) {
>     $permission->content = $this->content;
> } else {
>     throw new InvalidArgumentException(get_class($permission) . ' must be instance of ' . AbstractContentPermission::class);
> }
> 
> return $permission->verify($this->getSubject());
> ```

Yes, good idea. 

> * I think better property name will be `AbstractContentPermission->model` or `AbstractContentPermission->record`, because it is a bit strange to keep the code `$this->content->content->id` inside the class.

Yes, it's confusing. Let's rename that to `$model` as suggested.

> * Should we implement new methods like `$model->permissions->can*()`? `$model` is `ContentActiveRecord`
> * Also I think it will be good to keep same property `permissions` for class `humhub\modules\content\models\Content` because in many places we have a code like `$content->canView()`, and it will be better to have `$content->permissions->canView()` instead of `$content->model->permissions->canView()`.

Hmmm. what do you think about the following:

- Methods like `$record->can*()` are deprecated.
- Methods like `$record->content->can*()` remain, but are only shortcuts for e.g. `$record->content>getPermissionManager($user)->can($record->viewPermission)`
- Otherwise no other new methods in PermissionManager `->permissions->can*'.

**Comment by luke-** [comment]
We should reconsider the following points of the current approach:

- There is a lot complexity added to the `PermissionManager` and makes `humhub\libs¸\BasePermission` deprecated which requires refactoring. Even if we can preserve backward compatibility for now.

- In `ContentActiveRecord` there are now a lot of attributes `editPermission`, `createPermission`, `managePermission`, `viewPermission`. Also it could be a bit confusing that some are `GroupPermission` and others are `ContentPermission`. 

- The methods signature of `AbstractPermissionManager::verify(BasePermission $permission)` although the `ContentPermissionManager::verify(ContentPermission $permission)` and the `GroupPermissionManager::verify(GroupPermission $p)` .

**Comment by yurabakhtin** [comment]
@luke-

> > * I think we could remove `$content` the first param from the method `AbstractContentPermission->verify(ContentActiveRecord $content, ?User $user)` and initialize this as property when we call this method from the `ContentPermissionManager->verify(BasePermission $permission)`, like code:
> > 
> > ```
> > if ($permission instanceof AbstractContentPermission) {
> >     $permission->content = $this->content;
> > } else {
> >     throw new InvalidArgumentException(get_class($permission) . ' must be instance of ' . AbstractContentPermission::class);
> > }
> > 
> > return $permission->verify($this->getSubject());
> > ```
> 
> Yes, good idea.
> 
> > * I think better property name will be `AbstractContentPermission->model` or `AbstractContentPermission->record`, because it is a bit strange to keep the code `$this->content->content->id` inside the class.
> 
> Yes, it's confusing. Let's rename that to `$model` as suggested.

Commit https://github.com/humhub/humhub/pull/6034/commits/b340db9881ac74c3f1429ffac49b9e976d1c88de.

**Comment by yurabakhtin** [comment]
@luke- 

> > * Should we implement new methods like `$model->permissions->can*()`? `$model` is `ContentActiveRecord`
> > * Also I think it will be good to keep same property `permissions` for class `humhub\modules\content\models\Content` because in many places we have a code like `$content->canView()`, and it will be better to have `$content->permissions->canView()` instead of `$content->model->permissions->canView()`.
> 
> Hmmm. what do you think about the following:
> 
> * Methods like `$record->can*()` are deprecated.

Ok, we can deprecate them. As I understand after that we should create new permissions for models from external modules, for example, code from the `humhub\modules\tasks\models\Task->canEdit()` will be moved into new `AbstractContentPermission`, and then the code https://github.com/humhub/humhub/pull/6034/files#diff-54e5adc6290d919d2f026ab9153c4503f43278a19a829c668e35f2b9682becb4R50-R54 will be deleted completely to avoid recursion:

```
        // Check if underlying models canEdit implementation
        // ToDo: Implement this as interface
        if (method_exists($this->model, 'canEdit') && $this->model->canEdit($user)) {
            return true;
        }
```

so my PR https://github.com/humhub/humhub/pull/5992 will be closed wthout merging.

> * Methods like `$record->content->can*()` remain, but are only shortcuts for e.g. `$record->content>getPermissionManager($user)->can($record->viewPermission)`

Ok.

> * Otherwise no other new methods in PermissionManager `->permissions->can*'.

Ok, if we keep `$record->content->can*()` then we don't need new `$record->permissions->can*()`.

**Comment by yurabakhtin** [comment]
@luke-

> We should reconsider the following points of the current approach:
> 
> * There is a lot complexity added to the `PermissionManager` and makes `humhub\libs\BasePermission` deprecated which requires refactoring. Even if we can preserve backward compatibility for now.

Should I start to use new `GroupPermission` in all external modules?

> * In `ContentActiveRecord` there are now a lot of attributes `editPermission`, `createPermission`, `managePermission`, `viewPermission`. Also it could be a bit confusing that some are `GroupPermission` and others are `ContentPermission`.

Yes, I have been thinking about issue two.
Currently we have only 2 new properties `$editPermission` and `$viewPermission`, but in future we may have more, like `$deletePermission`, `$movePermission` and etc.
I think we should move them all into the new `ContentPermissionManager`, and declare a new property `public $permissionManager = ContentPermissionManager::class`, so the `ContentActiveRecord` will be more clear.

> * The methods signature of `AbstractPermissionManager::verify(BasePermission $permission)` although the `ContentPermissionManager::verify(ContentPermission $permission)` and the `GroupPermissionManager::verify(GroupPermission $p)` .

Sorry, not clear what you mean here.
